### PR TITLE
Refactor/sites

### PR DESCRIPTION
**Pull request by alexanderleegs**
## Overview

This PR introduces the sites flow refactor. It introduces a newly created SitesRouter and SitesService, as well as updating the existing GithubService to have an additional checkHasAccess method.

**Breaking Changes**
- No - this PR is backwards compatible

**Features**:

Sites router
- GET / - retrieve list of all sites
- GET /:siteName - check if user has access
- GET /:siteName/lastUpdated - retrieve date of last update
- GET /:siteName/stagingUrl - retrieve staging url

**Notes**:
The SitesService uses both the existing GithubService as well as calling axios itself - this is due to the existing GithubService being tailored specifically for the sites endpoint. I'm open to discussion of whether to initialise another axios instance inside the GithubService specifically to deal with this endpoint, but I currently set it without as this is the only github endpoint with the `orgs/repo` prefix.

### CODE REVIEW

File: services/utilServices/SitesService.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
+const axios = require("axios")
+const Bluebird = require("bluebird")
+const _ = require("lodash")
+
+// Import error
```
Comment (seaerchin) [inline services/utilServices/SitesService.js:5]: nit: delete wrong comment

File: services/utilServices/SitesService.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    // Simultaneously retrieve all isomerpages repos
+    const paramsArr = []
+    for (let i = 0; i < ISOMERPAGES_REPO_PAGE_COUNT; i += 1) {
+      paramsArr.push({
+        per_page: GH_MAX_REPO_COUNT,
+        sort: "full_name",
+        page: i + 1,
+      })
+    }
```
Comment (seaerchin) [inline services/utilServices/SitesService.js:60]: wdyt about writing this as a map?

```js 
const paramsArr = _.fill(Array(ISOMERPAGES_REPO_PAGE_COUNT), null).map((_, idx) => ({
  per_page: GH_MAX_REPO_COUNT,
  sort: "full_name",
  page: i + 1,
}))```

File: services/utilServices/SitesService.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    const sites = await Bluebird.map(paramsArr, async (params) => {
+      const resp = await axios.get(endpoint, {
+        params,
+        headers: {
+          Authorization: `token ${accessToken}`,
+          "Content-Type": "application/json",
+        },
```
Comment (seaerchin) [inline services/utilServices/SitesService.js:68]: i think i've seen this repeated a ton and it merits its own helper method. might be out of scope (and not that important yet), so maybe we could file an issue?

File: services/utilServices/SitesService.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      })
+
+      return resp.data
+        .map((repoData) => {
+          const {
+            updated_at: updatedAt,
+            permissions,
+            name,
+            private: isPrivate,
+          } = repoData
```
Comment (seaerchin) [inline services/utilServices/SitesService.js:78]: wdyt about destructuring here?

File: services/utilServices/SitesService.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        })
+        .filter(
+          (repoData) =>
+            repoData.permissions.push === true &&
+            !ISOMER_ADMIN_REPOS.includes(repoData.repoName)
+        )
+    })
+
+    const flattenedSites = _.flatten(sites)
+    return flattenedSites
```
Comment (seaerchin) [inline services/utilServices/SitesService.js:95]: nit: could we inline this?

File: services/utilServices/SitesService.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    const numDaysAgo = Math.floor(gapInUpdate / (1000 * 60 * 60 * 24))
+    // return a message for number of days ago repo was last updated
+    switch (numDaysAgo) {
+      case 0:
+        return "Updated today"
+      case 1:
+        return "Updated 1 day ago"
+      default:
+        return `Updated ${numDaysAgo} days ago`
+    }
```
Comment (seaerchin) [inline services/utilServices/SitesService.js:46]: i'm not sure that this should be done on the backend - i'm of hte opinion that the backend returns purely data and makes no judgement on how that data is consumed. this limits flexibility on the FE side (eg: if another FE component calls the same endpoint but wishes to make a `Updated last week` statement) and loses information.

File: services/utilServices/SitesService.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    let stagingUrl
+
+    if (description) {
+      // Retrieve the url from the description - repo descriptions have varying formats, so we look for the first link
+      const descTokens = description.replace("/;/g", " ").split(" ")
+      // Staging urls also contain staging in their url
+      stagingUrl = descTokens.find(
+        (token) => token.includes("http") && token.includes("staging")
+      )
+    }
```
Comment (seaerchin) [inline services/utilServices/SitesService.js:125]: This implies that if users were to change their description, we could potentially have a misbehaving branch! In that case, would it be sensible to log the necessary details here so that we can identify the branch?

File: services/utilServices/SitesService.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    return stagingUrl
```
Comment (seaerchin) [inline services/utilServices/SitesService.js:127]: this implies that `stagingUrl` can be `undefined`, in which case we should adopt some preventive actions rather than doing nothing. perhaps a good backup here might be to direct them to the 404 page?

File: services/utilServices/__tests__/SitesService.spec.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    it("Checks when site was last updated", async () => {
+      mockGithubService.getRepoInfo.mockResolvedValue(repoInfo)
+      await expect(service.getLastUpdated(reqDetails)).resolves.toEqual(
+        `Updated ${numDaysAgo} days ago`
+      )
+      expect(mockGithubService.getRepoInfo).toHaveBeenCalledWith(reqDetails)
+    })
+  })
+
+  describe("getStagingUrl", () => {
```
Comment (seaerchin) [inline services/utilServices/__tests__/SitesService.spec.js:108]: could we add a test case when both config and description lack staging url?

**Review by seaerchin — COMMENTED** [review]
seems fine overall, just some minor comments wrt style/odd edge cases in our codebase!

File: newroutes/__tests__/Sites.spec.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+// Can't set request fields - will always be undefined
+const userId = undefined
+const accessToken = undefined
+
+const siteName = "siteName"
+
+const reqDetails = { siteName, accessToken }
+
+describe("Unlinked Pages Router", () => {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply newroutes/__tests__/Sites.spec.js:20]: https://github.com/isomerpages/isomercms-backend/pull/341/commits/8fe141dafd56c2d52886d11a9b89512a492c1b23

File: newroutes/sites.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+  async getSites(req, res) {
+    const { accessToken } = req
+    const siteNames = await this.sitesService.getSites({ accessToken })
+    return res.status(200).json({ siteNames })
+  }
+
+  async checkHasAccess(req, res) {
+    const { accessToken, userId } = req
+    const { siteName } = req.params
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply newroutes/sites.js]: https://github.com/isomerpages/isomercms-backend/pull/341/commits/132b081696e46097d6ff14ef30e06596530df4d2

File: newroutes/sites.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    const lastUpdated = await this.sitesService.getLastUpdated({
+      accessToken,
+      siteName,
+    })
+    return res.status(200).json({ lastUpdated })
+  }
+
+  async getStagingUrl(req, res) {
+    const { accessToken } = req
+    const { siteName } = req.params
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply newroutes/sites.js]: https://github.com/isomerpages/isomercms-backend/pull/341/commits/132b081696e46097d6ff14ef30e06596530df4d2

File: services/utilServices/SitesService.js (hunk 5 above)
Comment (alexanderleegs) [reply services/utilServices/SitesService.js]: https://github.com/isomerpages/isomercms-backend/pull/341/commits/e64781021882ea0289420e5ee8d68a4015f14341

File: services/utilServices/SitesService.js (hunk 2 above)
Comment (alexanderleegs) [reply services/utilServices/SitesService.js]: Yeah that's probably better, https://github.com/isomerpages/isomercms-backend/pull/341/commits/e64781021882ea0289420e5ee8d68a4015f14341

File: newroutes/sites.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+    await this.sitesService.checkHasAccess(
+      {
+        accessToken,
+        siteName,
+      },
+      { userId }
+    )
+    return res.status(200).send("OK")
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply newroutes/sites.js]: https://github.com/isomerpages/isomercms-backend/pull/341/commits/e64781021882ea0289420e5ee8d68a4015f14341

File: services/utilServices/SitesService.js (hunk 4 above)
Comment (alexanderleegs) [reply services/utilServices/SitesService.js]: f3b0110

File: services/utilServices/SitesService.js (hunk 7 above)
Comment (alexanderleegs) [reply services/utilServices/SitesService.js]: Great suggestion, I've made it throw an error if no staging url is found: 6d645cf

File: services/utilServices/SitesService.js (hunk 8 above)
Comment (alexanderleegs) [reply services/utilServices/SitesService.js]: See above, 6d645cf

File: services/utilServices/__tests__/SitesService.spec.js (hunk 9 above)
Comment (alexanderleegs) [reply services/utilServices/__tests__/SitesService.spec.js:99]: Added in 6d645cf!

File: services/utilServices/SitesService.js (hunk 6 above)
Comment (alexanderleegs) [reply services/utilServices/SitesService.js]: I think that makes a lot of sense, I've changed it in 66c43c8 and made an issue on the frontend so we don't forget when refactoring sites

File: services/utilServices/SitesService.js (hunk 1 above)
Comment (alexanderleegs) [reply services/utilServices/SitesService.js]: e647810

File: services/utilServices/SitesService.js (hunk 3 above)
Comment (alexanderleegs) [reply services/utilServices/SitesService.js:52]: This is used only in `SitesService`, `AuthService` and `GithubService` actually, since this is only relevant when querying Github - I'm wondering if the solution is to create a more general misc service to query endpoints which don't start with `https://api.github.com/repos/`? It felt out of place for these 2 services to have axios calls when the only other place which uses axios is `GithubService`

File: newmiddleware/auth.js [hunk 14]
```diff
@@ -255,6 +255,12 @@ class AuthMiddleware {
     auth.delete("/v2/auth/logout", this.noVerify)
     auth.get("/v2/auth/whoami", this.whoamiAuth)
 
+    // Sites
+    auth.get("/v2/sites", this.verifyJwt)
+    auth.get("/v2/sites/:siteName", this.verifyJwt)
+    auth.get("/v2/sites/:siteName/lastUpdated", this.verifyJwt)
+    auth.get("/v2/sites/:siteName/stagingUrl", this.verifyJwt)
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply newmiddleware/auth.js]: https://github.com/isomerpages/isomercms-backend/pull/341/commits/4a921fe26e9364b9096fe910a55f3467633f986f

File: services/utilServices/SitesService.js (hunk 3 above)
Comment (seaerchin) [reply services/utilServices/SitesService.js:52]: actually you're right - maybe we should just defer the methods that require headers to GHA and make our main services agnostic of the data provider (github) wdyt?

**Review by seaerchin — COMMENTED** [review]
looks good and will approve once the test comment is fixed! as an aside, i think destructuring should take place on the params level but this isn't important as we do destructure later on

File: services/utilServices/SitesService.js (hunk 3 above)
Comment (alexanderleegs) [reply services/utilServices/SitesService.js:52]: Yup i think that makes sense - i'll file an issue to tackle this in a future PR

**Review by seaerchin — APPROVED** [review]
minor test comments but lgtm otherwise! just go ahead n merge when the comment is fixed! 

as an aside, i think we shouldn't rebase _previous_ commits as prior review comments on that commit hash will be lost. also, you might wanna double check if you lost work as the prior nit about not returning `"Ok"` got reverted after being fixed (idt there's anything major though but just err on the side of safety)

File: newroutes/sites.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply newroutes/sites.js:36]: ^this commit got lost, unsure if there are others. i'm fine iwth not doing this as it's relatively minor

File: services/utilServices/__tests__/SitesService.spec.js [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+      }))
+      axios.get.mockImplementationOnce(() => ({
+        data: [],
+      }))
+
+      await expect(service.getSites({ accessToken })).resolves.toMatchObject(
+        expectedResp
+      )
+
+      expect(axios.get).toHaveBeenCalledTimes(3)
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply services/utilServices/__tests__/SitesService.spec.js:70]: Going to leave it as is for now - we'll refactor this out into a general axios call in a future PR!

File: services/utilServices/__tests__/SitesService.spec.js [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+  })
+
+  describe("getStagingUrl", () => {
+    const stagingUrl = "https://repo-staging.netlify.app"
+    it("Retrieves the staging url for a site from config if available", async () => {
+      mockConfigYmlService.read.mockResolvedValue({
+        content: {
+          staging: stagingUrl,
+        },
+      })
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply services/utilServices/__tests__/SitesService.spec.js:106]: https://github.com/isomerpages/isomercms-backend/pull/341/commits/fb05904b7799e50b5a89a5cce4323773ea49105d

File: newroutes/sites.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply newroutes/sites.js:36]: The commit was preserved as 9b05ff7 - rebasing seems to overwrite the original commit hash! I did reintroduce the `.send("OK")` in a later commit because one of the tests was failing, and elected to use `.send("OK")` to keep it in line with how we return responses in the rest of our routers
